frontend/src/metabase/dashboard/containers/DashboardApp.unit.spec.jsx

```diff
@@ -0,0 +1,121 @@
+import React from "react";
+import { Route } from "react-router";
+import { screen } from "@testing-library/react";
+
+import userEvent from "@testing-library/user-event";
+import fetchMock from "fetch-mock";
+import { renderWithProviders, waitForElementToBeRemoved } from "__support__/ui";
+import DashboardApp from "metabase/dashboard/containers/DashboardApp";
+import {
+  createMockCollection,
+  createMockDashboard,
+  createMockUser,
+} from "metabase-types/api/mocks";
+import {
+  setupCollectionItemsEndpoint,
+  setupCollectionsEndpoints,
+  setupDashboardEndpoints,
+} from "__support__/server-mocks";
+import { createMockDashboardState } from "metabase-types/store/mocks";
+import { BEFORE_UNLOAD_UNSAVED_MESSAGE } from "metabase/hooks/use-before-unload";
+
+const TEST_DASHBOARD = createMockDashboard({
+  id: 1,
+  name: "Example",
+});
+
+const TEST_COLLECTION = createMockCollection({
+  id: "root",
+});
+
+async function setup(user = createMockUser()) {
+  setupDashboardEndpoints(createMockDashboard(TEST_DASHBOARD));
+  setupCollectionsEndpoints([]);
+  setupCollectionItemsEndpoint(TEST_COLLECTION);
+
+  fetchMock.get("path:/api/bookmark", []);
+
+  const { mockEvent, events } = createMockEventListener();
+
+  const DashboardAppContainer = props => {
+    return (
+      <main>
+        <DashboardApp {...props} />
+      </main>
+    );
+  };
+
+  renderWithProviders(
+    <Route path="/dashboard/:slug" component={DashboardAppContainer} />,
+    {
+      initialRoute: `/dashboard/${TEST_DASHBOARD.id}`,
+      currentUser: user,
+      withRouter: true,
+      storeInitialState: {
+        dashboard: createMockDashboardState(),
+      },
+    },
+  );
+
+  await waitForElementToBeRemoved(() =>
+    screen.queryAllByTestId("loading-spinner"),
+  );
+
+  return { mockEvent, events };
+}
+
+function createMockEventListener() {
+  const events = {};
```

Comment:
Maybe we could simplify the setup.

```
const setup = (...) => {
  const eventMock = { preventDefault: jest.fn() });
  const eventListenerMock = jest.spyOn(window, "addEventListener");

  return { eventMock, eventListenerMock };
});
```

and the test:

```
const { eventMock, eventListenerMock } = await setup(); 

...

eventListenerMock.mock.calls
  .filter(([event]) => event === "beforeunload"))
  .forEach(([_, callback]) => callback(mockEvent));

expect(eventMock.preventDefault).toHaveBeenCalled();
expect(eventMock.returnValue).toBe(...);
```

Reply:
It looks good, I think the only thing I'm concerned about is reusability - the reason I put it in this function in the first place was with the thinking that we could use this in other files